### ISSUE TITLE
#[must_use] fires when generic type is unit type

### ISSUE THREAD

**Issue by wmmc88** [rust-lang/rust#141802]
I tried this code:

```rust
use core::convert::Infallible;

pub trait UnwrapInfallible<T> {
    #[must_use]
    fn unwrap_infallible(self) -> T;
}

impl<T> UnwrapInfallible<T> for Result<T, Infallible> {
    #[inline]
    fn unwrap_infallible(self) -> T {
        match self {
            Ok(value) => value,
        }
    }
}

fn main() {
    let result: Result<(), Infallible> = Ok(());
    let result2: Result<u32, Infallible> = Ok(1);
    
    result.unwrap_infallible();
    let _u = result2.unwrap_infallible();
}```

I expected to see this happen: no warnings
Instead, this happened: warning from unused `()` from `result.unwrap_infallible()`

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.86.0 (05f9846f8 2025-03-31)
binary: rustc
commit-hash: 05f9846f893b09a1be1fc8560e33fc3c815cfecb
commit-date: 2025-03-31
host: x86_64-pc-windows-msvc
release: 1.86.0
LLVM version: 19.1.7
```
It's also in latest nightly: `2025-5-29`

[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=e5fe35e6315162835a877736f4148fb6)

**Comment by Noratrieb** [comment]
If you call a function with the `#[must_use]` attribute, its return value is unconditionally must_use. This is expected.
